### PR TITLE
Allow image resources to reference custom cluster builders

### PR DESCRIPTION
**Pull request by athornton2012**
Addresses #203

### CODE REVIEW

**Comment by pivotal-issuemaster** [comment]
@athornton2012 Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/pivotal?repositoryId=pivotal/kpack&pullRequestId=230)!

[Click here](https://cla.pivotal.io/sync/pivotal?repositoryId=pivotal/kpack&pullRequestId=230) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

**Comment by pivotal-issuemaster** [comment]
@athornton2012 Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/pivotal?repositoryId=pivotal/kpack&pullRequestId=230)!

File: pkg/apis/build/v1alpha1/image_types.go [hunk 1]
```diff
@@ -37,15 +37,15 @@ type Image struct {
...
-	Builder                  ImageBuilder         `json:"builder,omitempty"`
-	ServiceAccount           string               `json:"serviceAccount,omitempty"`
-	Source                   SourceConfig         `json:"source"`
-	CacheSize                *resource.Quantity   `json:"cacheSize,omitempty"`
-	FailedBuildHistoryLimit  *int64               `json:"failedBuildHistoryLimit,omitempty"`
-	SuccessBuildHistoryLimit *int64               `json:"successBuildHistoryLimit,omitempty"`
-	ImageTaggingStrategy     ImageTaggingStrategy `json:"imageTaggingStrategy,omitempty"`
-	Build                    *ImageBuild          `json:"build,omitempty"`
+	Tag                      string                 `json:"tag"`
+	Builder                  corev1.ObjectReference `json:"builder,omitempty"`
```
Comment (matthewmcnew) [inline pkg/apis/build/v1alpha1/image_types.go]: Love it :)

File: pkg/reconciler/v1alpha1/clusterbuilder/clusterbuilder_test.go [hunk 2]
```diff
@@ -66,7 +66,7 @@ func testClusterBuilderReconciler(t *testing.T, when spec.G, it spec.S) {
 			Generation: initialGeneration,
 		},
 		Spec: v1alpha1.BuilderSpec{
-			Image: clusterImageName,
+			Image:        clusterImageName,
```
Comment (matthewmcnew) [inline pkg/reconciler/v1alpha1/clusterbuilder/clusterbuilder_test.go]: I wonder if we should start more agressively `go fmt ./..`

Comment (ashwin-venkatesh) [reply pkg/reconciler/v1alpha1/clusterbuilder/clusterbuilder_test.go]: I agree 100%

File: pkg/reconciler/v1alpha1/customclusterbuilder/customclusterbuilder.go [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+		}
+
+		return controller.NewPermanentError(creationError)
+	}
+
+	customBuilder.Status.BuilderRecord(builderRecord)
+	return c.updateStatus(customBuilder)
+}
+
+func (c *Reconciler) reconcileCustomBuilder(customBuilder *experimentalV1alpha1.CustomClusterBuilder) (v1alpha1.BuilderRecord, error) {
```
Comment (ashwin-venkatesh) [inline pkg/reconciler/v1alpha1/customclusterbuilder/customclusterbuilder.go]: can we rename this method to `reconcileBuilder`?

File: pkg/reconciler/v1alpha1/customclusterbuilder/customclusterbuilder.go [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+	KeychainFactory            registry.KeychainFactory
+}
+
+func (c *Reconciler) Reconcile(ctx context.Context, key string) error {
+	_, builderName, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	customBuilder, err := c.CustomClusterBuilderLister.Get(builderName)
```
Comment (ashwin-venkatesh) [inline pkg/reconciler/v1alpha1/customclusterbuilder/customclusterbuilder.go]: rename variable to just `builder`?

File: pkg/reconciler/v1alpha1/image/image.go [hunk 5]
```diff
@@ -83,26 +77,16 @@ func NewController(
...
-		(&v1alpha1.ClusterBuilder{}).GetGroupVersionKind(),
-	)))
+	duckbuilderInformer.AddEventHandler(reconciler.Handler(c.Tracker.OnChanged))
 
 	return impl
 }
 
 type Reconciler struct {
 	Client               versioned.Interface
+	DuckBuilderLister    *duckbuilder.DuckBuilderLister
```
Comment (ashwin-venkatesh) [inline pkg/reconciler/v1alpha1/image/image.go]: ❤️

File: pkg/reconciler/v1alpha1/image/image.go [hunk 6]
```diff
@@ -134,7 +118,7 @@ func (c *Reconciler) Reconcile(ctx context.Context, key string) error {
 }
 
 func (c *Reconciler) reconcileImage(image *v1alpha1.Image) (*v1alpha1.Image, error) {
-	builder, err := c.getBuilder(image)
+	builder, err := c.DuckBuilderLister.Namespace(image.Namespace).Get(image.Spec.Builder)
```
Comment (ashwin-venkatesh) [inline pkg/reconciler/v1alpha1/image/image.go]: This is beautiful!

**Pull request merged**